yapapi/strategy.py

```diff
@@ -168,33 +157,37 @@ async def score_offer(
 class DecreaseScoreForUnconfirmedAgreement(MarketStrategy):
     """A market strategy that modifies a base strategy based on history of agreements."""
 
-    base_strategy: MarketStrategy
-    factor: float
-
-    def __init__(self, base_strategy, factor):
+    def __init__(self, base_strategy: MarketStrategy, engine: "_Engine", factor: float):
         """
         :param base_strategy: the base strategy around which this strategy is wrapped
+        :param engine: the main Golem engine, access point to the agreement history
         :param factor: the factor by which the score of an offer for a provider which
                        failed to confirm the last agreement proposed to them will be multiplied
         """
         self.base_strategy = base_strategy
+        self.engine = engine
         self.factor = factor
         self._logger = logging.getLogger(f"{__name__}.{type(self).__name__}")
 
     async def decorate_demand(self, demand: DemandBuilder) -> None:
         """Decorate `demand` using the base strategy."""
         await self.base_strategy.decorate_demand(demand)
 
-    async def score_offer(
-        self, offer: rest.market.OfferProposal, history: Optional[ComputationHistory] = None
-    ) -> float:
+    async def score_offer(self, offer: rest.market.OfferProposal) -> float:
         """Score `offer` using the base strategy and apply penalty if needed.
 
         If the offer issuer failed to approve the previous agreement (if any)
         then the base score is multiplied by `self._factor`.
         """
         score = await self.base_strategy.score_offer(offer)
-        if history and history.rejected_last_agreement(offer.issuer) and score > 0:
+        if score > 0 and self.rejected_last_agreement(offer.issuer):
             self._logger.debug("Decreasing score for offer %s from '%s'", offer.id, offer.issuer)
             score *= self.factor
         return score
+
+    def rejected_last_agreement(self, provider_id: str) -> bool:
+        """Current implementation checks only if provider rejected last agreement during the job with
```

Comment:
okay, the implementation here is the main reason I have an issue with this pull request ... 

with the current (previous) implementation, we're scoring offers using a record that has a defined, explicitly-defined interface ... 

the record _happens_ to be an "AgreementsPool" but _doesn't_ _have_ _to_ _be_ - it's just an interface (`ComputationHistory`) that we chose "AgreementsPool" to implement...

[ and btw because of the above, I believe it's _improper_ to refer to it as "AgreementsPool" in any code that uses it (like you did in your example that you showed on the status call) because it doesn't have to be an agreements pool -> it only has to implement that interface ]

After your proposed change, we're sacrificing the explicit interface for a vague implementation that takes the whole `engine` ... I don't think this is a change for better API-wise ... 

Especially that you're _also_ performing a check here using an assumption that's not guaranteed to be true (that the job id stays the same during the whole execution of a task - which we may not want to be true if multiple tasks are executed simultaneously) ...

Reply:
> okay, the implementation here is the main reason I have an issue with this pull request ...
> 
> with the current (previous) implementation, we're scoring offers using a record that has a defined, explicitly-defined interface ...
> 
> the record _happens_ to be an "AgreementsPool" but _doesn't_ _have_ _to_ _be_ - it's just an interface (`ComputationHistory`) that we chose "AgreementsPool" to implement...
> 
> [ and btw because of the above, I believe it's _improper_ to refer to it as "AgreementsPool" in any code that uses it (like you did in your example that you showed on the status call) because it doesn't have to be an agreements pool -> it only has to implement that interface ]

I don't understand this.
In the current implementation, `AgreementsPool` is passed to `score_offer` via `yapapi` internals and developer has **no** way of changing this without messing with `yapapi.engine.Job`. So - unless I'm missing something - it **has** to be `AgreementsPool` in any `yapapi`-related context (of course, you can implement this interface some other way, but still `yapapi` will be passing instance of `AgreementsPool`).

> 
> After your proposed change, we're sacrificing the explicit interface for a vague implementation that takes the whole `engine` ... I don't think this is a change for better API-wise ...

I think it's worse for `DecreaseScoreForUnconfirmedStrategy` and better for other strategies.
And also - all the ugliness will disappear with https://github.com/golemfactory/yapapi/issues/747.


> 
> Especially that you're _also_ performing a check here using an assumption that's not guaranteed to be true (that the job id stays the same during the whole execution of a task - which we may not want to be true if multiple tasks are executed simultaneously) ...

That's true and I agree that this is a serious problem.

All in all I still think it's a change for the better (and I have no idea how to deal with problems you mentioned in a way other than https://github.com/golemfactory/yapapi/issues/747, without deep internal changes).